Allow usage of safe intrinsics outside unsafe blocks

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.
-->

Some intrinsics, such as `rotate_left` or `wrapping_add`, are safe and do not require the usage of an unsafe block.

I tried this code:

```rust
extern "rust-intrinsic" {
    pub fn rotate_left<T>(l: T, r: T) -> T;
}

fn main() -> i32 {
    let a = 15;
    let b = 15;

    let c = rotate_left(a, b);

    0
}
```

I expected to see this happen: no error

Equivalent proper Rust code:

```rust
#![feature(core_intrinsics)]

use std::intrinsics::rotate_left;

fn main() {
    let a = 15;
    let b = 15;

    let c = rotate_left(a, b);
}
```

Instead, this happened: 

```rust
test.rs:9:13: error: call to extern function requires unsafe function or block
    9 |     let c = rotate_left(a, b);
      |             ^
```

I didn't have much time to look into this and couldn't really find where this is handled. But here is a backtrace that produces this error. I would've gone through it next if I had more time:

<details>
<summary>unsafe intrinsic backtrace</summary>

```
error[E0133]: call to unsafe function is unsafe and requires unsafe function or block
 --> main.rs:8:13
  |
8 |     let c = cttz_nonzero(a);
  |             ^^^^^^^^^^^^^^^ call to unsafe function
  |
  = note: consult the function's documentation for information on how to avoid undefined behavior

thread 'rustc' panicked at 'aborting due to `-Z treat-err-as-bug=1`', compiler/rustc_errors/src/lib.rs:1450:27
stack backtrace:
   0:     0x7faca78feb44 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hf8137a2d0ec3eb55
   1:     0x7faca7981f8e - core::fmt::write::he0cdbcb1b4f9137e
   2:     0x7faca791fb11 - std::io::Write::write_fmt::hdb60c8913756632e
   3:     0x7faca78fe999 - std::sys_common::backtrace::print::hd1bf38bd5ba01a8f
   4:     0x7faca7927807 - std::panicking::default_hook::{{closure}}::hdd6e4b37dd7c82c0
   5:     0x7faca7927633 - std::panicking::default_hook::h1e7367f0ea885543
   6:     0x7faca81cd614 - <alloc[65971c2eb6ab290e]::boxed::Box<dyn for<'a, 'b> core[5571ea7c97e175b]::ops::function::Fn<(&'a core[5571ea7c97e175b]::panic::panic_info::PanicInfo<'b>,), Output = ()> + core[5571ea7c97e175b]::marker::Send + core[5571ea7c97e175b]::marker::Sync> as core[5571ea7c97e175b]::ops::function::Fn<(&core[5571ea7c97e175b]::panic::panic_info::PanicInfo,)>>::call
                               at /home/pkrones/rust-lang/rust/library/alloc/src/boxed.rs:1949:9
   7:     0x7faca81cd614 - rustc_driver[f4d8627d82ee58e8]::DEFAULT_HOOK::{closure#0}::{closure#0}
                               at /home/pkrones/rust-lang/rust/compiler/rustc_driver/src/lib.rs:1152:13
   8:     0x7faca7927c08 - std::panicking::rust_panic_with_hook::h045fd0ac3760a96e
   9:     0x7faca7927fc9 - std::panicking::begin_panic_handler::{{closure}}::hcff7c428a6b06052
  10:     0x7faca7927f44 - std::sys_common::backtrace::__rust_end_short_backtrace::hdea9d242d3c43e08
  11:     0x7faca79278d2 - rust_begin_unwind
  12:     0x7faca78d4083 - core::panicking::panic_fmt::hb8c3375d42574ff4
  13:     0x7facaa622a63 - <rustc_errors[bf65e971d66277a9]::HandlerInner>::emit_diagnostic
  14:     0x7facaa61f710 - <rustc_errors[bf65e971d66277a9]::Handler>::emit_diagnostic
                               at /home/pkrones/rust-lang/rust/compiler/rustc_errors/src/lib.rs:1020:9
  15:     0x7facaa61f710 - <rustc_errors[bf65e971d66277a9]::ErrorGuaranteed as rustc_errors[bf65e971d66277a9]::diagnostic_builder::EmissionGuarantee>::diagnostic_builder_emit_producing_guarantee
                               at /home/pkrones/rust-lang/rust/compiler/rustc_errors/src/diagnostic_builder.rs:139:28
  16:     0x7faca85b4041 - <rustc_errors[bf65e971d66277a9]::diagnostic_builder::DiagnosticBuilder<rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>::emit
                               at /home/pkrones/rust-lang/rust/compiler/rustc_errors/src/diagnostic_builder.rs:293:9
  17:     0x7faca85b4041 - rustc_mir_transform[179bf7ccc75f482e]::check_unsafety::check_unsafety
                               at /home/pkrones/rust-lang/rust/compiler/rustc_mir_transform/src/check_unsafety.rs:583:17
  18:     0x7faca8303913 - rustc_interface[e4067e6ed6429006]::passes::analysis::{closure#3}
                               at /home/pkrones/rust-lang/rust/compiler/rustc_interface/src/passes.rs:931:17
  19:     0x7faca8303913 - <rustc_data_structures[ad78b4262079ceb]::profiling::VerboseTimingGuard>::run::<(), rustc_interface[e4067e6ed6429006]::passes::analysis::{closure#3}>
                               at /home/pkrones/rust-lang/rust/compiler/rustc_data_structures/src/profiling.rs:739:9
  20:     0x7faca8303913 - <rustc_session[a9461aaf65cd7d33]::session::Session>::time::<(), rustc_interface[e4067e6ed6429006]::passes::analysis::{closure#3}>
                               at /home/pkrones/rust-lang/rust/compiler/rustc_session/src/utils.rs:10:9
  21:     0x7faca82ebc42 - rustc_interface[e4067e6ed6429006]::passes::analysis
                               at /home/pkrones/rust-lang/rust/compiler/rustc_interface/src/passes.rs:927:5
  22:     0x7faca9676ab4 - <rustc_query_system[c152b0b76cc57d03]::query::config::QueryVTable<rustc_query_impl[50dff0e644211e9e]::plumbing::QueryCtxt, (), core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>>::compute
                               at /home/pkrones/rust-lang/rust/compiler/rustc_query_system/src/query/config.rs:43:9
  23:     0x7faca9676ab4 - rustc_query_system[c152b0b76cc57d03]::query::plumbing::execute_job::<rustc_query_impl[50dff0e644211e9e]::plumbing::QueryCtxt, (), core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>::{closure#0}
                               at /home/pkrones/rust-lang/rust/compiler/rustc_query_system/src/query/plumbing.rs:384:55
  24:     0x7faca9676ab4 - stacker[c7f5accb384a487d]::maybe_grow::<core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>, rustc_query_system[c152b0b76cc57d03]::query::plumbing::execute_job<rustc_query_impl[50dff0e644211e9e]::plumbing::QueryCtxt, (), core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>::{closure#0}>
                               at /home/pkrones/.cargo/registry/src/github.com-1ecc6299db9ec823/stacker-0.1.14/src/lib.rs:55:9
  25:     0x7faca9676ab4 - rustc_data_structures[ad78b4262079ceb]::stack::ensure_sufficient_stack::<core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>, rustc_query_system[c152b0b76cc57d03]::query::plumbing::execute_job<rustc_query_impl[50dff0e644211e9e]::plumbing::QueryCtxt, (), core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>::{closure#0}>
                               at /home/pkrones/rust-lang/rust/compiler/rustc_data_structures/src/stack.rs:17:5
  26:     0x7faca9676ab4 - <rustc_query_impl[50dff0e644211e9e]::plumbing::QueryCtxt as rustc_query_system[c152b0b76cc57d03]::query::QueryContext>::start_query::<core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>, rustc_query_system[c152b0b76cc57d03]::query::plumbing::execute_job<rustc_query_impl[50dff0e644211e9e]::plumbing::QueryCtxt, (), core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>::{closure#0}>::{closure#0}::{closure#0}
                               at /home/pkrones/rust-lang/rust/compiler/rustc_query_impl/src/plumbing.rs:112:17
  27:     0x7faca9676ab4 - rustc_middle[251fe8b5feeb05c7]::ty::context::tls::enter_context::<<rustc_query_impl[50dff0e644211e9e]::plumbing::QueryCtxt as rustc_query_system[c152b0b76cc57d03]::query::QueryContext>::start_query<core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>, rustc_query_system[c152b0b76cc57d03]::query::plumbing::execute_job<rustc_query_impl[50dff0e644211e9e]::plumbing::QueryCtxt, (), core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>::{closure#0}>::{closure#0}::{closure#0}, core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>::{closure#0}
                               at /home/pkrones/rust-lang/rust/compiler/rustc_middle/src/ty/context.rs:1929:50
  28:     0x7faca9676ab4 - rustc_middle[251fe8b5feeb05c7]::ty::context::tls::set_tlv::<rustc_middle[251fe8b5feeb05c7]::ty::context::tls::enter_context<<rustc_query_impl[50dff0e644211e9e]::plumbing::QueryCtxt as rustc_query_system[c152b0b76cc57d03]::query::QueryContext>::start_query<core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>, rustc_query_system[c152b0b76cc57d03]::query::plumbing::execute_job<rustc_query_impl[50dff0e644211e9e]::plumbing::QueryCtxt, (), core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>::{closure#0}>::{closure#0}::{closure#0}, core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>::{closure#0}, core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>
                               at /home/pkrones/rust-lang/rust/compiler/rustc_middle/src/ty/context.rs:1913:9
  29:     0x7faca9676ab4 - rustc_middle[251fe8b5feeb05c7]::ty::context::tls::enter_context::<<rustc_query_impl[50dff0e644211e9e]::plumbing::QueryCtxt as rustc_query_system[c152b0b76cc57d03]::query::QueryContext>::start_query<core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>, rustc_query_system[c152b0b76cc57d03]::query::plumbing::execute_job<rustc_query_impl[50dff0e644211e9e]::plumbing::QueryCtxt, (), core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>::{closure#0}>::{closure#0}::{closure#0}, core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>
                               at /home/pkrones/rust-lang/rust/compiler/rustc_middle/src/ty/context.rs:1929:9
  30:     0x7faca9676ab4 - <rustc_query_impl[50dff0e644211e9e]::plumbing::QueryCtxt as rustc_query_system[c152b0b76cc57d03]::query::QueryContext>::start_query::<core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>, rustc_query_system[c152b0b76cc57d03]::query::plumbing::execute_job<rustc_query_impl[50dff0e644211e9e]::plumbing::QueryCtxt, (), core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>::{closure#0}>::{closure#0}
                               at /home/pkrones/rust-lang/rust/compiler/rustc_query_impl/src/plumbing.rs:111:13
  31:     0x7faca9676ab4 - rustc_middle[251fe8b5feeb05c7]::ty::context::tls::with_related_context::<<rustc_query_impl[50dff0e644211e9e]::plumbing::QueryCtxt as rustc_query_system[c152b0b76cc57d03]::query::QueryContext>::start_query<core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>, rustc_query_system[c152b0b76cc57d03]::query::plumbing::execute_job<rustc_query_impl[50dff0e644211e9e]::plumbing::QueryCtxt, (), core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>::{closure#0}>::{closure#0}, core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>::{closure#0}
                               at /home/pkrones/rust-lang/rust/compiler/rustc_middle/src/ty/context.rs:1973:13
  32:     0x7faca9676ab4 - rustc_middle[251fe8b5feeb05c7]::ty::context::tls::with_context::<rustc_middle[251fe8b5feeb05c7]::ty::context::tls::with_related_context<<rustc_query_impl[50dff0e644211e9e]::plumbing::QueryCtxt as rustc_query_system[c152b0b76cc57d03]::query::QueryContext>::start_query<core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>, rustc_query_system[c152b0b76cc57d03]::query::plumbing::execute_job<rustc_query_impl[50dff0e644211e9e]::plumbing::QueryCtxt, (), core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>::{closure#0}>::{closure#0}, core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>::{closure#0}, core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>::{closure#0}
                               at /home/pkrones/rust-lang/rust/compiler/rustc_middle/src/ty/context.rs:1957:40
  33:     0x7faca9676ab4 - rustc_middle[251fe8b5feeb05c7]::ty::context::tls::with_context_opt::<rustc_middle[251fe8b5feeb05c7]::ty::context::tls::with_context<rustc_middle[251fe8b5feeb05c7]::ty::context::tls::with_related_context<<rustc_query_impl[50dff0e644211e9e]::plumbing::QueryCtxt as rustc_query_system[c152b0b76cc57d03]::query::QueryContext>::start_query<core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>, rustc_query_system[c152b0b76cc57d03]::query::plumbing::execute_job<rustc_query_impl[50dff0e644211e9e]::plumbing::QueryCtxt, (), core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>::{closure#0}>::{closure#0}, core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>::{closure#0}, core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>::{closure#0}, core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>
                               at /home/pkrones/rust-lang/rust/compiler/rustc_middle/src/ty/context.rs:1946:22
  34:     0x7faca9676ab4 - rustc_middle[251fe8b5feeb05c7]::ty::context::tls::with_context::<rustc_middle[251fe8b5feeb05c7]::ty::context::tls::with_related_context<<rustc_query_impl[50dff0e644211e9e]::plumbing::QueryCtxt as rustc_query_system[c152b0b76cc57d03]::query::QueryContext>::start_query<core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>, rustc_query_system[c152b0b76cc57d03]::query::plumbing::execute_job<rustc_query_impl[50dff0e644211e9e]::plumbing::QueryCtxt, (), core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>::{closure#0}>::{closure#0}, core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>::{closure#0}, core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>
                               at /home/pkrones/rust-lang/rust/compiler/rustc_middle/src/ty/context.rs:1957:9
  35:     0x7faca9676ab4 - rustc_middle[251fe8b5feeb05c7]::ty::context::tls::with_related_context::<<rustc_query_impl[50dff0e644211e9e]::plumbing::QueryCtxt as rustc_query_system[c152b0b76cc57d03]::query::QueryContext>::start_query<core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>, rustc_query_system[c152b0b76cc57d03]::query::plumbing::execute_job<rustc_query_impl[50dff0e644211e9e]::plumbing::QueryCtxt, (), core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>::{closure#0}>::{closure#0}, core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>
                               at /home/pkrones/rust-lang/rust/compiler/rustc_middle/src/ty/context.rs:1970:9
  36:     0x7faca9676ab4 - <rustc_query_impl[50dff0e644211e9e]::plumbing::QueryCtxt as rustc_query_system[c152b0b76cc57d03]::query::QueryContext>::start_query::<core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>, rustc_query_system[c152b0b76cc57d03]::query::plumbing::execute_job<rustc_query_impl[50dff0e644211e9e]::plumbing::QueryCtxt, (), core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>::{closure#0}>
                               at /home/pkrones/rust-lang/rust/compiler/rustc_query_impl/src/plumbing.rs:100:9
  37:     0x7faca9676ab4 - rustc_query_system[c152b0b76cc57d03]::query::plumbing::execute_job::<rustc_query_impl[50dff0e644211e9e]::plumbing::QueryCtxt, (), core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>
                               at /home/pkrones/rust-lang/rust/compiler/rustc_query_system/src/query/plumbing.rs:384:22
  38:     0x7faca9676ab4 - rustc_query_system[c152b0b76cc57d03]::query::plumbing::try_execute_query::<rustc_query_impl[50dff0e644211e9e]::plumbing::QueryCtxt, rustc_query_system[c152b0b76cc57d03]::query::caches::DefaultCache<(), core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>>
                               at /home/pkrones/rust-lang/rust/compiler/rustc_query_system/src/query/plumbing.rs:343:44
  39:     0x7faca9739c32 - rustc_query_system[c152b0b76cc57d03]::query::plumbing::get_query::<rustc_query_impl[50dff0e644211e9e]::queries::analysis, rustc_query_impl[50dff0e644211e9e]::plumbing::QueryCtxt>
                               at /home/pkrones/rust-lang/rust/compiler/rustc_query_system/src/query/plumbing.rs:702:36
  40:     0x7faca952af62 - <rustc_query_impl[50dff0e644211e9e]::Queries as rustc_middle[251fe8b5feeb05c7]::ty::query::QueryEngine>::analysis
                               at /home/pkrones/rust-lang/rust/compiler/rustc_query_impl/src/lib.rs:56:1
  41:     0x7faca81bb4b4 - <rustc_middle[251fe8b5feeb05c7]::ty::query::TyCtxtAt>::analysis
                               at /home/pkrones/rust-lang/rust/compiler/rustc_middle/src/ty/query.rs:259:17
  42:     0x7faca81bb4b4 - <rustc_middle[251fe8b5feeb05c7]::ty::context::TyCtxt>::analysis
                               at /home/pkrones/rust-lang/rust/compiler/rustc_middle/src/ty/query.rs:240:17
  43:     0x7faca81bb4b4 - rustc_driver[f4d8627d82ee58e8]::run_compiler::{closure#1}::{closure#2}::{closure#3}
                               at /home/pkrones/rust-lang/rust/compiler/rustc_driver/src/lib.rs:378:30
  44:     0x7faca81bb4b4 - <rustc_interface[e4067e6ed6429006]::passes::QueryContext>::enter::<rustc_driver[f4d8627d82ee58e8]::run_compiler::{closure#1}::{closure#2}::{closure#3}, core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>::{closure#0}
                               at /home/pkrones/rust-lang/rust/compiler/rustc_interface/src/passes.rs:787:42
  45:     0x7faca81bb4b4 - rustc_middle[251fe8b5feeb05c7]::ty::context::tls::enter_context::<<rustc_interface[e4067e6ed6429006]::passes::QueryContext>::enter<rustc_driver[f4d8627d82ee58e8]::run_compiler::{closure#1}::{closure#2}::{closure#3}, core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>::{closure#0}, core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>::{closure#0}
                               at /home/pkrones/rust-lang/rust/compiler/rustc_middle/src/ty/context.rs:1929:50
  46:     0x7faca81bb4b4 - rustc_middle[251fe8b5feeb05c7]::ty::context::tls::set_tlv::<rustc_middle[251fe8b5feeb05c7]::ty::context::tls::enter_context<<rustc_interface[e4067e6ed6429006]::passes::QueryContext>::enter<rustc_driver[f4d8627d82ee58e8]::run_compiler::{closure#1}::{closure#2}::{closure#3}, core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>::{closure#0}, core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>::{closure#0}, core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>
                               at /home/pkrones/rust-lang/rust/compiler/rustc_middle/src/ty/context.rs:1913:9
  47:     0x7faca81bb4b4 - rustc_middle[251fe8b5feeb05c7]::ty::context::tls::enter_context::<<rustc_interface[e4067e6ed6429006]::passes::QueryContext>::enter<rustc_driver[f4d8627d82ee58e8]::run_compiler::{closure#1}::{closure#2}::{closure#3}, core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>::{closure#0}, core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>
                               at /home/pkrones/rust-lang/rust/compiler/rustc_middle/src/ty/context.rs:1929:9
  48:     0x7faca81bb4b4 - <rustc_interface[e4067e6ed6429006]::passes::QueryContext>::enter::<rustc_driver[f4d8627d82ee58e8]::run_compiler::{closure#1}::{closure#2}::{closure#3}, core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>
                               at /home/pkrones/rust-lang/rust/compiler/rustc_interface/src/passes.rs:787:9
  49:     0x7faca81bcd28 - rustc_driver[f4d8627d82ee58e8]::run_compiler::{closure#1}::{closure#2}
                               at /home/pkrones/rust-lang/rust/compiler/rustc_driver/src/lib.rs:377:13
  50:     0x7faca81bcd28 - <rustc_interface[e4067e6ed6429006]::interface::Compiler>::enter::<rustc_driver[f4d8627d82ee58e8]::run_compiler::{closure#1}::{closure#2}, core[5571ea7c97e175b]::result::Result<core[5571ea7c97e175b]::option::Option<rustc_interface[e4067e6ed6429006]::queries::Linker>, rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>
                               at /home/pkrones/rust-lang/rust/compiler/rustc_interface/src/queries.rs:381:19
  51:     0x7faca81add57 - rustc_driver[f4d8627d82ee58e8]::run_compiler::{closure#1}
                               at /home/pkrones/rust-lang/rust/compiler/rustc_driver/src/lib.rs:309:22
  52:     0x7faca81add57 - rustc_interface[e4067e6ed6429006]::interface::create_compiler_and_run::<core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>, rustc_driver[f4d8627d82ee58e8]::run_compiler::{closure#1}>::{closure#1}
                               at /home/pkrones/rust-lang/rust/compiler/rustc_interface/src/interface.rs:323:13
  53:     0x7faca81add57 - rustc_span[adee44cace36fa3b]::with_source_map::<core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>, rustc_interface[e4067e6ed6429006]::interface::create_compiler_and_run<core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>, rustc_driver[f4d8627d82ee58e8]::run_compiler::{closure#1}>::{closure#1}>
                               at /home/pkrones/rust-lang/rust/compiler/rustc_span/src/lib.rs:986:5
  54:     0x7faca81bd8a3 - rustc_interface[e4067e6ed6429006]::interface::create_compiler_and_run::<core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>, rustc_driver[f4d8627d82ee58e8]::run_compiler::{closure#1}>
                               at /home/pkrones/rust-lang/rust/compiler/rustc_interface/src/interface.rs:317:5
  55:     0x7faca81b6432 - rustc_interface[e4067e6ed6429006]::interface::run_compiler::<core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>, rustc_driver[f4d8627d82ee58e8]::run_compiler::{closure#1}>::{closure#0}
                               at /home/pkrones/rust-lang/rust/compiler/rustc_interface/src/interface.rs:339:12
  56:     0x7faca81b6432 - <scoped_tls[f3befed452337256]::ScopedKey<rustc_span[adee44cace36fa3b]::SessionGlobals>>::set::<rustc_interface[e4067e6ed6429006]::interface::run_compiler<core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>, rustc_driver[f4d8627d82ee58e8]::run_compiler::{closure#1}>::{closure#0}, core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>
                               at /home/pkrones/.cargo/registry/src/github.com-1ecc6299db9ec823/scoped-tls-1.0.0/src/lib.rs:137:9
  57:     0x7faca81b4609 - rustc_span[adee44cace36fa3b]::create_session_globals_then::<core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>, rustc_interface[e4067e6ed6429006]::interface::run_compiler<core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>, rustc_driver[f4d8627d82ee58e8]::run_compiler::{closure#1}>::{closure#0}>
                               at /home/pkrones/rust-lang/rust/compiler/rustc_span/src/lib.rs:112:5
  58:     0x7faca81b4609 - rustc_interface[e4067e6ed6429006]::util::run_in_thread_pool_with_globals::<rustc_interface[e4067e6ed6429006]::interface::run_compiler<core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>, rustc_driver[f4d8627d82ee58e8]::run_compiler::{closure#1}>::{closure#0}, core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>::{closure#0}
                               at /home/pkrones/rust-lang/rust/compiler/rustc_interface/src/util.rs:159:32
  59:     0x7faca81b4609 - std[814b3ec6f1be451f]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[e4067e6ed6429006]::util::run_in_thread_pool_with_globals<rustc_interface[e4067e6ed6429006]::interface::run_compiler<core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>, rustc_driver[f4d8627d82ee58e8]::run_compiler::{closure#1}>::{closure#0}, core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>::{closure#0}, core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>
                               at /home/pkrones/rust-lang/rust/library/std/src/sys_common/backtrace.rs:122:18
  60:     0x7faca81bf4ee - <std[814b3ec6f1be451f]::thread::Builder>::spawn_unchecked_::<rustc_interface[e4067e6ed6429006]::util::run_in_thread_pool_with_globals<rustc_interface[e4067e6ed6429006]::interface::run_compiler<core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>, rustc_driver[f4d8627d82ee58e8]::run_compiler::{closure#1}>::{closure#0}, core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>::{closure#0}, core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>::{closure#1}::{closure#0}
                               at /home/pkrones/rust-lang/rust/library/std/src/thread/mod.rs:505:17
  61:     0x7faca81bf4ee - <core[5571ea7c97e175b]::panic::unwind_safe::AssertUnwindSafe<<std[814b3ec6f1be451f]::thread::Builder>::spawn_unchecked_<rustc_interface[e4067e6ed6429006]::util::run_in_thread_pool_with_globals<rustc_interface[e4067e6ed6429006]::interface::run_compiler<core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>, rustc_driver[f4d8627d82ee58e8]::run_compiler::{closure#1}>::{closure#0}, core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>::{closure#0}, core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>::{closure#1}::{closure#0}> as core[5571ea7c97e175b]::ops::function::FnOnce<()>>::call_once
                               at /home/pkrones/rust-lang/rust/library/core/src/panic/unwind_safe.rs:271:9
  62:     0x7faca81bf4ee - std[814b3ec6f1be451f]::panicking::try::do_call::<core[5571ea7c97e175b]::panic::unwind_safe::AssertUnwindSafe<<std[814b3ec6f1be451f]::thread::Builder>::spawn_unchecked_<rustc_interface[e4067e6ed6429006]::util::run_in_thread_pool_with_globals<rustc_interface[e4067e6ed6429006]::interface::run_compiler<core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>, rustc_driver[f4d8627d82ee58e8]::run_compiler::{closure#1}>::{closure#0}, core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>::{closure#0}, core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>::{closure#1}::{closure#0}>, core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>
                               at /home/pkrones/rust-lang/rust/library/std/src/panicking.rs:492:40
  63:     0x7faca81bf4ee - std[814b3ec6f1be451f]::panicking::try::<core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>, core[5571ea7c97e175b]::panic::unwind_safe::AssertUnwindSafe<<std[814b3ec6f1be451f]::thread::Builder>::spawn_unchecked_<rustc_interface[e4067e6ed6429006]::util::run_in_thread_pool_with_globals<rustc_interface[e4067e6ed6429006]::interface::run_compiler<core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>, rustc_driver[f4d8627d82ee58e8]::run_compiler::{closure#1}>::{closure#0}, core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>::{closure#0}, core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>::{closure#1}::{closure#0}>>
                               at /home/pkrones/rust-lang/rust/library/std/src/panicking.rs:456:19
  64:     0x7faca81b3f80 - std[814b3ec6f1be451f]::panic::catch_unwind::<core[5571ea7c97e175b]::panic::unwind_safe::AssertUnwindSafe<<std[814b3ec6f1be451f]::thread::Builder>::spawn_unchecked_<rustc_interface[e4067e6ed6429006]::util::run_in_thread_pool_with_globals<rustc_interface[e4067e6ed6429006]::interface::run_compiler<core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>, rustc_driver[f4d8627d82ee58e8]::run_compiler::{closure#1}>::{closure#0}, core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>::{closure#0}, core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>::{closure#1}::{closure#0}>, core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>
                               at /home/pkrones/rust-lang/rust/library/std/src/panic.rs:137:14
  65:     0x7faca81b3f80 - <std[814b3ec6f1be451f]::thread::Builder>::spawn_unchecked_::<rustc_interface[e4067e6ed6429006]::util::run_in_thread_pool_with_globals<rustc_interface[e4067e6ed6429006]::interface::run_compiler<core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>, rustc_driver[f4d8627d82ee58e8]::run_compiler::{closure#1}>::{closure#0}, core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>::{closure#0}, core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>::{closure#1}
                               at /home/pkrones/rust-lang/rust/library/std/src/thread/mod.rs:504:30
  66:     0x7faca81b3f80 - <<std[814b3ec6f1be451f]::thread::Builder>::spawn_unchecked_<rustc_interface[e4067e6ed6429006]::util::run_in_thread_pool_with_globals<rustc_interface[e4067e6ed6429006]::interface::run_compiler<core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>, rustc_driver[f4d8627d82ee58e8]::run_compiler::{closure#1}>::{closure#0}, core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>::{closure#0}, core[5571ea7c97e175b]::result::Result<(), rustc_errors[bf65e971d66277a9]::ErrorGuaranteed>>::{closure#1} as core[5571ea7c97e175b]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
                               at /home/pkrones/rust-lang/rust/library/core/src/ops/function.rs:248:5
  67:     0x7faca7907448 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h156a36301f1cba83
  68:     0x7faca79098f7 - std::sys::unix::thread::Thread::new::thread_start::he42b8935706ec0ef
  69:     0x7faca768c54d - <unknown>
  70:     0x7faca7711874 - clone
  71:                0x0 - <unknown>

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.65.0-dev running on x86_64-unknown-linux-gnu

note: compiler flags: -Z treat-err-as-bug

query stack during panic:
#0 [analysis] running analysis passes on this crate
end of query stack
```
</details>

--- a/gcc/rust/checks/errors/rust-unsafe-checker.cc
+++ b/gcc/rust/checks/errors/rust-unsafe-checker.cc
@@ -90,6 +90,50 @@ check_unsafe_call (HIR::Function *fn, Location locus, const std::string &kind)
 		   kind.c_str ());
 }
 
+static bool
+is_safe_intrinsic (const std::string &fn_name)
+{
+  static const std::unordered_set<std::string> safe_intrinsics = {
+    "abort",
+    "size_of",
+    "min_align_of",
+    "needs_drop",
+    "caller_location",
+    "add_with_overflow",
+    "sub_with_overflow",
+    "mul_with_overflow",
+    "wrapping_add",
+    "wrapping_sub",
+    "wrapping_mul",
+    "saturating_add",
+    "saturating_sub",
+    "rotate_left",
+    "rotate_right",
+    "ctpop",
+    "ctlz",
+    "cttz",
+    "bswap",
+    "bitreverse",
+    "discriminant_value",
+    "type_id",
+    "likely",
+    "unlikely",
+    "ptr_guaranteed_eq",
+    "ptr_guaranteed_ne",
+    "minnumf32",
+    "minnumf64",
+    "maxnumf32",
+    "rustc_peek",
+    "maxnumf64",
+    "type_name",
+    "forget",
+    "black_box",
+    "variant_count",
+  };
+
+  return safe_intrinsics.find (fn_name) != safe_intrinsics.end ();
+}
+
 static void
 check_extern_call (HIR::ExternalItem *maybe_fn, HIR::ExternBlock *parent_block,
 		   Location locus)
@@ -103,9 +147,16 @@ check_extern_call (HIR::ExternalItem *maybe_fn, HIR::ExternBlock *parent_block,
   // hand, any function defined in a block with a specific ABI (even `extern
   // "Rust"` blocks) is unsafe to call
 
-  if (maybe_fn->get_extern_kind () == ExternalItem::ExternKind::Function)
-    rust_error_at (locus,
-		   "call to extern function requires unsafe function or block");
+  if (maybe_fn->get_extern_kind () != ExternalItem::ExternKind::Function)
+    return;
+
+  // Some intrinsics are safe to call
+  if (parent_block->get_abi () == Rust::ABI::INTRINSIC
+      && is_safe_intrinsic (maybe_fn->get_item_name ()))
+    return;
+
+  rust_error_at (locus,
+		 "call to extern function requires unsafe function or block");
 }
 
 void

--- a/gcc/testsuite/rust/compile/unsafe10.rs
+++ b/gcc/testsuite/rust/compile/unsafe10.rs
@@ -1,0 +1,12 @@
+extern "rust-intrinsic" {
+    pub fn rotate_left<T>(l: T, r: T) -> T;
+}
+
+fn main() -> i32 {
+    let a = 15;
+    let b = 15;
+
+    let _ = rotate_left(a, b);
+
+    0
+}

